libopae++/samples/hello_fpga-3.cpp

```diff
@@ -0,0 +1,121 @@
+#include "opae/cxx/buffers.h"
+#include <uuid/uuid.h>
+#include <algorithm>
+#include <iostream>
+#include <chrono>
+#include <thread>
+
+using namespace opae::fpga::resources;
+
+static const char* NLB0_AFUID                      = "D8424DC4-A4A3-C413-F89E-433683F9040B";
+static const uint64_t CL                           = 64;
+static const uint64_t KB                           = 1024;
+static const uint64_t MB                           = KB * 1024;
+static const uint64_t GB                           = MB * 1024;
+static const uint64_t LOG2_CL                      = 6;
+static const size_t LPBK1_DSM_SIZE                 = 4*KB;
+static const size_t LPBK1_BUFFER_SIZE              = 1*KB;
+static const size_t LPBK1_BUFFER_ALLOCATION_SIZE   = 4*KB;
+static const uint64_t CSR_SRC_ADDR                 = 0x0120;
+static const uint32_t CSR_DST_ADDR                 = 0x0128;
+static const uint32_t CSR_CTL                      = 0x0138;
+static const uint32_t CSR_CFG                      = 0x0140;
+static const uint32_t CSR_NUM_LINES                = 0x0130;
+static const uint32_t DSM_STATUS_TEST_COMPLETE     = 0x40;
+static const uint64_t CSR_AFU_DSM_BASEL            = 0x0110;
+static const uint64_t CSR_AFU_DSM_BASEH            = 0x0114;
+
+static inline uint64_t cacheline_aligned_addr(uint64_t num) {
+  return num >> LOG2_CL;
+}
+
+int main(int argc, char* argv[]) {
+  // look for accelerator with NLB0_AFUID
+  properties filter;
+  filter.guid.parse(NLB0_AFUID);
+  // Call the accelerator::run static function with the following args:
+  // 1. The filter used to locate the accelerator
+  // 2. The flags used when opening the accelerator
+  // 3. The size of the initial buffer - more can be requested from accel
+  // 4. The callback called once an accelerator is found and opened
+  //    This callback takes as arguments the accelerator and the buffer allocated
+  return accelerator::run(filter,                        
+                          FPGA_OPEN_SHARED,
+                          1*GB,
+                          [](accelerator::ptr_t accel, buffer_sice::ptr_t buffer){
+    // allocate buffers
+    auto buffers = buffer->split({ LPBK1_DSM_SIZE,
+                                   LPBK1_BUFFER_ALLOCATION_SIZE,
+                                   LPBK1_BUFFER_ALLOCATION_SIZE });
+    auto dsm = buffers[0];
+    auto inp = buffers[1];
+    auto out = buffers[2];
+
+    // initialize buffers
+    std::fill_n(dsm->get(), LPBK1_DSM_SIZE, 0);
+    std::fill_n(inp->get(), LPBK1_BUFFER_SIZE, 0xAF);
+    std::fill_n(out->get(), LPBK1_BUFFER_SIZE, 0xBE);
+
+    accel->write(CSR_AFU_DSM_BASEL, dsm->iova());
+    accel->configure<uint32_t>({ 
+      {CSR_CTL, static_cast<uint32_t>(0)},
+      {CSR_CTL, static_cast<uint32_t>(1)}
+    });
+    accel->configure<uint64_t>({
+      {CSR_SRC_ADDR, cacheline_aligned_addr(inp->iova())},
+      {CSR_DST_ADDR, cacheline_aligned_addr(out->iova())},
+      {CSR_NUM_LINES, LPBK1_BUFFER_SIZE / 1 * CL}
+    });
```

Comment:
I don't know; this assumes IO space is only used for configuration data - AFUs may do anything with that (i.e. map it to local memory). Also "configure" is already overloaded (as in "reconfigure the FPGA with a new bitstream"), which may lead to confusion.